### PR TITLE
Fastmath: Added required version

### PR DESCRIPTION
**Pull request by tldahlgren**
I'm working on "build system" documentation for ``BundlePackage`` and noticed I could not build this package because it does not have a version.

This commit adds a version placeholder and tweaks the URL (since the redirect caused problems for me) .

### CODE REVIEW

File: var/spack/repos/builtin/packages/fastmath/package.py [hunk 1]
```diff
@@ -13,7 +13,9 @@ class Fastmath(BundlePackage):
     libraries for ODE's, Time Integrators, Iterative, Non-Linear, and Direct
     Solvers."""
 
-    homepage = "www.fastmath-scidac.org/"
+    homepage = "https://fastmath-scidac.org/"
+
+    version('0.0.1')  # TODO: What should this be?
```
Comment (markcmiller86) [inline var/spack/repos/builtin/packages/fastmath/package.py]: AFAIK, FASTMath package suite was itself never versioned. Maybe @balay can comment further.

**Comment by adamjstewart** [comment]
Why is a version and URL required for a BundlePackage?

Let's say I want a BundlePackage called `autotools` that contains:
```python
depends_on('automake')
depends_on('autoconf')
depends_on('libtool')
depends_on('m4')
```
I should be able to `spack install autotools` and have it install the latest version of these 4 packages. A version of `autotools` does not need to exist (and doesn't really make sense).

**Comment by tgamblin** [comment]
Technically not required, but I’m pretty sure we assume in many places right now that something concrete will always have a version.  I don’t think this would be super hard to change, but it’s also not a huge priority given the other stuff we’re on the hook for right now.

**Comment by markcmiller86** [comment]
> I should be able to `spack install autotools` and have it install the latest version of these 4 packages. A version of `autotools` does not need to exist (and doesn't really make sense).

While there may be no argument to the above, it is definitely true that FASTMath and similar *suites* of inter-related libraries and associated APIs do indeed need to have a version number because it is almost *never* the case that all the individual member package versions can vary while the whole remains fully functional. The *suite's* version number would then indicate the combination of member package numbers that are known to work together. So, what we should really see here, for a given suite version number is a bunch of version-specific depends on statements for the individual member packages. We don't have that here now, but maybe someone on FASTMath does now (@balay might know for sure).

**Comment by balay** [comment]
@markcmiller86 I'm not sure what you are looking for here.. what you are describing here is xsdk - but from prior discussion - fastmath metapacakge wanted to exist separately and different than xsdk.

And if fastmath metapackage won't have versions - [and always refer to the latest version of member packages] - you can use some generic version name here..

```

diff --git a/var/spack/repos/builtin/packages/fastmath/package.py b/var/spack/repos/builtin/packages/fastmath/package.py
index 463197cd9..33aed28ca 100644
--- a/var/spack/repos/builtin/packages/fastmath/package.py
+++ b/var/spack/repos/builtin/packages/fastmath/package.py
@@ -15,6 +15,8 @@ class Fastmath(BundlePackage):
 
     homepage = "www.fastmath-scidac.org/"
 
+    version('latest')
+
     depends_on('amrex')  # default is 3 dimensions
     depends_on('chombo@3.2')
     depends_on('hypre~internal-superlu')
```

**Comment by tldahlgren** [comment]
IIRC I was not able to aluminate the version directed requirement from the package.

I’ll have to look through my notes if I can find more information to share

On Dec 4, 2019, at 7:27 PM, Adam J. Stewart <notifications@github.com> wrote:

﻿

Why is a version and URL required for a BundlePackage?

Let's say I want a BundlePackage called autotools that contains:

depends_on('automake')
depends_on('autoconf')
depends_on('libtool')
depends_on('m4')

I should be able to spack install autotools and have it install the latest version of these 4 packages. A version of autotools does not need to exist (and doesn't really make sense).

**Comment by tldahlgren** [comment]
> Why is a version and URL required for a BundlePackage?
> 
> Let's say I want a BundlePackage called `autotools` that contains:
> 
> ```python
> depends_on('automake')
> depends_on('autoconf')
> depends_on('libtool')
> depends_on('m4')
> ```
> 
> I should be able to `spack install autotools` and have it install the latest version of these 4 packages. A version of `autotools` does not need to exist (and doesn't really make sense).

IIRC I was not able to aluminate the version directed requirement from the package.

I’ll have to look through my notes if I can find more information to share.

**Comment by tldahlgren** [comment]
> > Why is a version and URL required for a BundlePackage?
> > Let's say I want a BundlePackage called `autotools` that contains:
> > ```python
> > depends_on('automake')
> > depends_on('autoconf')
> > depends_on('libtool')
> > depends_on('m4')
> > ```
> > 
> > 
> > I should be able to `spack install autotools` and have it install the latest version of these 4 packages. A version of `autotools` does not need to exist (and doesn't really make sense).
> 
> IIRC I was not able to eliminate the version directive requirement from the package.
> 
> I’ll have to look through my notes to see if I can find more information to share.

Todd's right about Spack assuming anything concretized will have a version.  The function getting tripped up here is ``concretize_version``.

**Pull request merged**